accepted/00000-http-api.md

```diff
@@ -0,0 +1,220 @@
+```
+
+On each request, the Spark route wrapper shall unpack both types of arguments in the request, deserialize the payload, and call the appropriate handler method with unpacked arguments.
+
+To provide a uniform approach to handler registration, Java annotations can be used to specify a method as an endpoint handler. These annotations can be interpreted during initialization to register all the handlers to both frameworks. As a side benefit, all the handler classes in a predetermined package tree can be automatically identified, without the need to add them in a list manually.
+
+Java annotations shall also be used to determine attributes of an endpoint, such as whether it is a read-only method or not.
+
+Below is an example of such an annotation structure:
+
+#### @ApiHandler("namespace") Target=Class
+
+A class marked with this annotation will be registered as the handler of the specified namespace in the XMLRPC framework. The class will further be inspected on initialization to discover any annotated handler methods. Found methods will be registered to Spark as route handlers.
+
+#### @XmlRpcApiEndpoint Target=Method
+#### @HttpApiEndpoint Target=Method
+
+Each of these annotations shall mark a method to act as a handler for one of the APIs.
+
+**HTTP:** These methods will be registered to Spark as route handlers during initialization.
+
+**XMLRPC:** `BaseHandler` will search for a requested method in a list of methods annotated with `@XmlRpcApiEndpoint` in the class during a request.
+
+In most cases, both annotations shall be used together. Exclusive use of any of them shall be reserved for edge cases where the endpoint is not applicable to the specific protocol.
+
+#### @ReadOnly Target=Method
+
+An endpoint handled by this method will be set as a read-only endpoint.
+
+**HTTP:** Spark will create `GET` endpoints for the methods marked with this annotation. If not used, Spark's method wrapper will check the requesting user and block access if restricted.
+
+**XMLRPC:** `BaseHandler` will check this annotation against the requesting user and block access if restricted.
+
+#### @Unauthorized Target=Method
+
+An endpoint annotated with `@Unauthorized` may be called with or without a logged in user. The handler method may or may not have a `User` parameter in the signature. If it has, the parameter may be `null`, in case the user is not logged in.
+
+#### @RouteParam Target=Parameter
+
+This annotation specifies a parameter to be represented with a route parameter, rather than a querystring parameter. `@RouteParam` may not be used for object types as they have to be transferred via the request body.
+
+### Example 1
+
+Consider the following XMLRPC handler method:
+```java
+public ActivationKey getDetails(User user, String key)
+```
+
+With the proposed annotations, it may be written as:
+```java
+@XmlRpcApiEndpoint
+@HttpApiEndpoint
+@ReadOnly
+public ActivationKey getDetails(User user, String key)
```

Comment:
The plan is to interpret the registered methods with the following algorithm:
1. ~Get all the `@RouteParam` annotated parameters from the base URL sections~
2. Get all the rest of the parameters of **primitive** type from the querystring
3. Get all the object parameters from the request body in JSON

I can already say from the comments that we can eliminate No.1 and use querystring parameters only.

The registration code will take care of the differentiation between the primitive and complex types and look for the values in the right place.

Consider this example:
```java
public myApiEndpoint(User user, int id, MyComplexObject myObject, Map<..> myMap)
```
A request to this endpoint will be registered in a way that it must be called in the following form:
```
POST .../myApiEndpoint?id=<ID>
{
  myObject: {/*MyComplexObject serialized in JSON*/},
  myMap: {/*Map serialized in JSON*/}
}
```

I've chosen this approach because it'll require a minimal amount of work with the existing API. However, there are some limitations with this:
1. You need to know how the above algorithm works to know how exactly your request should look like. Hopefully, we'll be able to reflect this in the doc generation
2. Primitive parameters will always go to querystring, complex types will always go to method body. I don't see any problems with this since all the existing methods can follow this just fine.
3. Since we also have the concrete distinction of GET=read-only, POST=write and complex types are put into the POST body, it won't be possible to add complex type params (maps, objects) to a readonly method. This also shouldn't be a problem since it doesn't make much sense to have complex parameters in any read-only methods (mostly list/lookup methods that need just an ID). I will double-check this with all the existing RO endpoints.

Another, most simplistic alternative would be to use `POST` and only `POST`, embedding all the parameters into the JSON body. This requires the least amount of work for the implementation, but in the end, we must decide from the users' perspective, and how the API looks from outside above all.

I'll try to rephrase all this info into the actual RFC as well.